ovirt_imageio/client/_upload.py

```diff
@@ -0,0 +1,152 @@
+
+    cmd.add_argument(
+        "--sparse",
+        action="store_true",
+        help="Create sparse disk. Cannot be used with raw format on "
+             "block storage.")
+
+    cmd.add_argument(
+        "--id",
+        type=_validate_uuid,
+        help="A UUID for the new disk. If not specified oVirt will "
+             "create a new UUID.")
+
+    cmd.add_argument(
+        "--disable-backup",
+        action="store_true",
+        help="Creates a disk that can not be used for incremental backup. "
+             "Mandatory with raw format, optional with qcow2 format.")
+
+    size = _options.Size(minimum=1, default=_api.MAX_WORKERS, maximum=8)
+    cmd.add_argument(
+        "--max-workers",
+        type=size,
+        default=size.default,
+        help=f"Maximum number of workers (range: {size.minimum}-"
+             f"{size.maximum}, default: {size.default}).")
+
+    size = _options.Size(
+        minimum=64 * KiB, default=_api.BUFFER_SIZE, maximum=16 * MiB)
+    cmd.add_argument(
+        "--buffer-size",
+        type=size,
+        default=size.default,
+        help=f"Buffer size per worker (range: {size.minimum}-"
+             f"{size.maximum}, default: {size.default}).")
+
+    cmd.add_argument(
+        "filename",
+        type=_validate_file,
+        help="Path to image (e.g. /path/to/image.raw). "
+             "Supported formats: raw, qcow2, iso.")
+
+
+def _validate_uuid(id):
+    """
+    Validate that the argument is a valid uuid.
+
+    Returns:
+        str: normalized UUID.
+    """
+    return str(uuid.UUID(id))
+
+
+def _validate_file(file):
+    """
+    Validate that the file exists.
+
+    Raises:
+        FileNotFoundError
+
+    Returns:
+        str: filename
+    """
+    if not os.path.exists(file):
+        raise ArgumentTypeError(f"{file} does not exist")
+    return file
+
+
+def _create_image_info(filename):
+    raw_info = qemu_img.info(filename)
+    if raw_info["format"] not in ("qcow2", "raw"):
+        raise RuntimeError(f"Unsupported image format {raw_info['format']}")
+
+    return _ovirt.ImageInfo(
+        raw_info["filename"], raw_info["format"], raw_info["virtual-size"])
+
+
+def upload_disk(args):
+    image_info = _create_image_info(args.filename)
+    con = _ovirt.connect(args)
+    with closing(con):
+        disk_format = args.format or image_info.format
+        image_measure = qemu_img.measure(image_info.filename, disk_format)
+        disk = _ovirt.add_disk(
+            con=con,
+            image_info=image_info,
+            initial_size=image_measure["required"],
+            sd_name=args.sd_name,
+            id=args.id,
+            sparse=args.sparse,
+            disable_backup=args.disable_backup,
+            format=_ovirt.RAW if disk_format == "raw" else _ovirt.COW)
+
+        host = _ovirt.find_host(con, args.sd_name)
+        transfer = _ovirt.create_transfer(con, disk, host=host)
+        try:
+            with _ui.ProgressBar() as pb:
```

Comment:
Lets create the progress at the top of the function, and update the phase during the function:

    with _ui.ProgressBar(phase="inspecting image") as pb:
         ...
         pb.phase = "creating disk"
         ...
         pb.phase = "creating transfer"
         ...
         pb.phase = "uploading image"
         ...
         ph.phase = "finalizing transfer"
         ...

We can also do this later, but creating the progress at the start is good change even if you don't add the phases yet.

Reply:
I'll do that, with phase changes included. Hopefully we have both patches ready and we can integrate them in order. Otherwise the phases are easy to remove before merging if needed.